zkvm/src/blockchain/block.rs

```diff
@@ -0,0 +1,121 @@
+    pub fn id(&self) -> BlockID {
+        let mut t = Transcript::new(b"ZkVM.blockheader");
+        t.commit_u64(b"version", self.version);
+        t.commit_u64(b"height", self.height);
+        t.commit_bytes(b"previd", &self.prev.0);
+        t.commit_u64(b"timestamp_ms", self.timestamp_ms);
+        t.commit_bytes(b"txroot", &self.txroot);
+        t.commit_bytes(b"utxoroot", &self.utxoroot);
+        t.commit_bytes(b"nonceroot", &self.nonceroot);
+        t.commit_u64(b"refscount", self.refscount);
+        t.commit_bytes(b"ext", &self.ext);
+
+        let mut result = [0u8; 32];
+        t.challenge_bytes(b"id", &mut result);
+        BlockID(result)
+    }
+
+    pub fn make_initial(timestamp_ms: u64, refscount: u64) -> BlockHeader {
+        BlockHeader {
+            version: 1,
+            height: 1,
+            prev: BlockID([0; 32]),
+            timestamp_ms: timestamp_ms,
+            txroot: [0; 32],
+            utxoroot: [0; 32],
+            nonceroot: [0; 32],
+            refscount: refscount,
+            ext: Box::new([]),
+        }
+    }
+}
+
+pub struct Block {
+    pub header: BlockHeader,
+    pub txs: Vec<Tx>,
+}
+
+impl Block {
+    pub fn validate(&self, prev: &BlockHeader) -> Result<Vec<TxLog>, BCError> {
+        if self.header.version < prev.version {
+            return Err(BCError::VersionReversion);
+        }
+        if self.header.version == 1 && self.header.ext.len() != 0 {
+            return Err(BCError::IllegalExtension);
+        }
+        if self.header.height != prev.height + 1 {
+            return Err(BCError::BadHeight);
+        }
+        if self.header.prev != prev.id() {
+            return Err(BCError::MismatchedPrev);
+        }
+        if self.header.timestamp_ms <= prev.timestamp_ms {
+            return Err(BCError::BadBlockTimestamp);
+        }
+        if self.header.refscount > prev.refscount + 1 {
+            return Err(BCError::BadRefscount);
+        }
+
+        let mut txlogs: Vec<TxLog> = Vec::new();
+        let mut txids: Vec<TxID> = Vec::new();
+
+        for tx in self.txs.iter() {
+            if tx.header.mintime_ms > self.header.timestamp_ms
+                || self.header.timestamp_ms > tx.header.maxtime_ms
+            {
+                return Err(BCError::BadTxTimestamp);
+            }
+            if self.header.version == 1 && tx.header.version != 1 {
+                return Err(BCError::BadTxVersion);
+            }
+
+            // TODO(bobg): The API currently requires anticipating how many multipliers (generators) are needed,
+            // _before_ knowing what's in the transaction.
+            // The value is 64 for each range proof,
+            // and there are at least as many range proofs as outputs in the tx.
+            // Guessing a value should _not_ be part of the tx-verifying API.
+            // Related: https://github.com/dalek-cryptography/bulletproofs/pull/263
+            let bp_gens = BulletproofGens::new(64 * 64, 1);
```

Comment:
Generators are expensive to compute and must be passed as an argument to whatever verification logic that wants to use them (as `&BulletproofGens`)

Reply:
Same thing - sounds like this is future work. Resolving.

Reply:
this one is actual suggestion for PR.

Reply:
(see other APIs that take &BulletproofGens as argument - r1cs::Verifier, zkvm::Verifier)

Reply:
Oh, are you suggesting that the call signature for Block::validate should include a `&BulletproofGens`? That wasn't clear.

Reply:
yeah. Sorry!